### PR TITLE
Add `conversation-activity-filter` feature

### PR DESCRIPTION
**Pull request by fregante**
Closes https://github.com/sindresorhus/refined-github/issues/3764
Closes https://github.com/sindresorhus/refined-github/issues/615
Replaces #3847

## Screenshot

(readme screenshot wip)

Screen Shot 1

Screen Shot 4

Screen Shot 3

## Test

Every event is on https://github.com/sindresorhus/refined-github/pull/4030
Some events are on https://github.com/sindresorhus/refined-github/pull/3847 but you can also test the native "X event hidden" there, which defers loading events in the middle

To test this feature, it's best to replace the CSS with:

```css
:root:root .rgh-conversation-timeline-filtered {
	/* display: none !important; */
	outline: solid 5px red;
}
```

So that the elements are highlighted instead of hidden (which is harder to keep track of, mentally)

Example of the above CSS in "Show only unresolved comments" state:

Screen Shot 7

## Wishlist (new PR?)

- [x] Move to header so it's always visible
- [x] Ensure that the 2 instances are synced
- [ ] Hide resolved review comments that are among unresolved comments: https://github.com/sindresorhus/refined-github/pull/4030#discussion_r584184425
- [/] ~~Add "Events hidden, click to undo" message at the end of the conversation~~ _Since it's always visible in the header now, this is no longer needed_

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#615: Collapse large sequences of label, milestone, etc changes](https://issuehunt.io/repos/51769689/issues/615)
---
</details>

### CODE REVIEW

**Comment by yakov116** [comment]
I will test tomorrow.

File: readme.md [hunk 1]
```diff
@@ -222,6 +223,7 @@ Thanks for contributing! 🦋🙌
 - [](# "git-checkout-pr") [Adds copy-pastable git commands to checkout a PR.](https://user-images.githubusercontent.com/16872793/96938908-8e024f80-1499-11eb-8976-0caf95175dd6.png)
 - [](# "jump-to-change-requested-comment") [Adds a link to jump to the latest changed requested comment.](https://user-images.githubusercontent.com/19198931/98718312-418b9f00-23c9-11eb-8da2-dfb616e95eb6.gif)
 
+
```
Comment (yakov116) [inline readme.md:226]:
```suggestion
```

File: source/features/conversation-timeline-filter.tsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+function init(): void {
+	observer.observe('#partial-users-participants', {
+		add: addToSidebar
+	});
+}
+
+function deinit(): void {
+	observer.disconnect();
+}
```
Comment (yakov116) [inline source/features/conversation-timeline-filter.tsx:166]: Can we just inline it? (line 174)

Comment (fregante) [reply source/features/conversation-timeline-filter.tsx:166]: No need, for the same reason why we don’t inline init

File: readme.md [hunk 3]
```diff
@@ -198,6 +198,7 @@ Thanks for contributing! 🦋🙌
 - [](# "clean-conversation-headers") [Removes duplicate information in the header of issues and PRs ("User wants to merge X commits from Y into Z")](https://user-images.githubusercontent.com/44045911/89736767-686ec800-da9e-11ea-81c3-252e9813140b.png)
 - [](# "dim-bots") [Dims commits and PRs by bots to reduce noise.](https://user-images.githubusercontent.com/1402241/65263190-44c52b00-db36-11e9-9b33-d275d3c8479d.gif)
 - [](# "esc-to-cancel") [Adds a shortcut to cancel editing a conversation title: <kbd>esc</kbd>.](https://user-images.githubusercontent.com/35100156/98303086-d81d2200-1fbd-11eb-8529-70d48d889bcf.gif)
+- [](# "conversation-timeline-filter") [Adds filters to hide unwanted things in conversations and adds ability to load all hidden items.](https://user-images.githubusercontent.com/3819892/103170243-2d014a00-4843-11eb-8a32-7e28eb755f00.gif)
```
Comment (fregante) [inline readme.md:201]:
```suggestion
- [](# "conversation-timeline-filter") [Lets you hide every event except comments or resolved comments in issues and PRs.](https://user-images.githubusercontent.com/3819892/103170243-2d014a00-4843-11eb-8a32-7e28eb755f00.gif)
```

**Comment by yakov116** [comment]
I tested! Looks amazing!

Not sure if you want me to test for a few days.

The only thing I was was that https://github.com/sindresorhus/refined-github/pull/4030#pullrequestreview-600213486 showed when I selected `Only show unresolved reviews`.

Also on this PR the "Show hidden items" is there due to the amount of commits. That were not hiding right?

**Comment by fregante** [comment]
> Not sure if you want me to test for a few days.

Yeah I don’t want to merge it for tomorrow, I want to see if moving it to the header is feasible 
 
> The only thing I was was that [#4030 (review)](https://github.com/sindresorhus/refined-github/pull/4030#pullrequestreview-600213486) showed when I selected `Only show unresolved reviews`.

The main comment of a review can’t be “resolved” so it will never be hidden. 

> Also on this PR the "Show hidden items" is there due to the amount of commits. That were not hiding right?

Yeah, that item might contain more unresolved comments but we don’t know until we open it. The original PR also auto-loaded it for this reason, but that part of the feature was rejected multiple times before.

File: source/features/conversation-timeline-filter.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+void features.add(__filebasename, {
+	include: [
+		pageDetect.isPRConversation,
+		pageDetect.isIssue
```
Comment (yakov116) [inline source/features/conversation-timeline-filter.tsx:169]: Why isIssue here?

Comment (fregante) [reply source/features/conversation-timeline-filter.tsx:169]: It works on issues too. If we have isConversation, it should use that

Comment (yakov116) [reply source/features/conversation-timeline-filter.tsx:169]: What are we hiding?

Comment (fregante) [reply source/features/conversation-timeline-filter.tsx:169]: Try it

**Comment by fregante** [comment]
I merged "Show only resolved reviews" into "Show only resolved comments". I didn't really make sense to "hide resolved comments but show resolved reviews."

This brings it down to 3 states:

- Off
- Hide events
- Hide events and resolved comments

This means it could probably be shown as 3 radios inline, without a header:

- [x] Show all
- [ ] Show comments only
- [ ] Show unresolved comments only

I'm still thinking of how this can be displayed in the header though, so the radios might not work there.

**Comment by fregante** [comment]
Moved to the header. This is now much cleaner and visible, and also doesn't lengthen the sidebar, which makes `sticky-conversation-sidebar` activate less often.

**Comment by fregante** [comment]
Done. This should now only need a readme update and reviews.

**Comment by yakov116** [comment]
🔥  Looking better by the moment!

**Comment by pradyunsg** [comment]
This is great! Thanks @fregante! ^.^

**Pull request merged**